compliance_controller/compliance_controller/src/compliance_controller.cpp

```diff
@@ -0,0 +1,315 @@
+                        "Unexpected error: No trajectory defined at current time. Please contact the package maintainer.");
+        return;
+      }
+
+      // Add compliance adjustments (velocity and position, integrated from velocity)
+      JointTrajectoryController::desired_joint_state_.velocity[0] = JointTrajectoryController::desired_joint_state_.velocity[0] + compliance_velocity_adjustment_.data[i];
+      
+      // Sometimes the control loop executes so fast, ROS time does not update
+      if ( time_data.uptime != last_trajectory_update_time_ )
+      {
+        last_trajectory_update_time_ = time_data.uptime;
+        integrated_traj_position_adjustment_[i] += compliance_velocity_adjustment_.data[i]*time_data.period.toSec();
+        JointTrajectoryController::desired_joint_state_.position[0] += integrated_traj_position_adjustment_[i];
+      }
+
+      JointTrajectoryController::desired_state_.position[i] = JointTrajectoryController::desired_joint_state_.position[0];
+      JointTrajectoryController::desired_state_.velocity[i] = JointTrajectoryController::desired_joint_state_.velocity[0];
+      JointTrajectoryController::desired_state_.acceleration[i] = JointTrajectoryController::desired_joint_state_.acceleration[0]; ;
+
+      JointTrajectoryController::state_joint_error_.position[0] = angles::shortest_angular_distance(JointTrajectoryController::current_state_.position[i], 
+        JointTrajectoryController::desired_joint_state_.position[0]);
+      JointTrajectoryController::state_joint_error_.velocity[0] = JointTrajectoryController::desired_joint_state_.velocity[0] - JointTrajectoryController::current_state_.velocity[i];
+      JointTrajectoryController::state_joint_error_.acceleration[0] = 0.0;
+
+      JointTrajectoryController::state_error_.position[i] = angles::shortest_angular_distance(JointTrajectoryController::current_state_.position[i],
+        JointTrajectoryController::desired_joint_state_.position[0]);
+      JointTrajectoryController::state_error_.velocity[i] = JointTrajectoryController::desired_joint_state_.velocity[0] - JointTrajectoryController::current_state_.velocity[i];
+      JointTrajectoryController::state_error_.acceleration[i] = 0.0;
+
+      //Check tolerances
+      const typename JointTrajectoryController::RealtimeGoalHandlePtr rt_segment_goal = segment_it->getGoalHandle();
+      if (rt_segment_goal && rt_segment_goal == JointTrajectoryController::rt_active_goal_)
+      {
+        // Check tolerances
+        if (time_data.uptime.toSec() < segment_it->endTime())
+        {
+          // Currently executing a segment: check path tolerances
+          const joint_trajectory_controller::SegmentTolerancesPerJoint< Scalar >& joint_tolerances = segment_it->getTolerances();
+
+          if (!checkStateTolerancePerJoint(JointTrajectoryController::state_joint_error_, joint_tolerances.state_tolerance))
+          {
+            if (JointTrajectoryController::verbose_)
```

Comment:
at this point your 5 indentations in, this function should be broken up into a helper function for, say, the inside of the for loop